### DISCUSSION TITLE
Does "Leveling active" checkbox need to be ON?

### DISCUSSION THREAD

**Discussion by shtefcs** [Jyers/Marlin#1212] [Q&A] [answered]
Hello, great work on new UI/UX and upgrades.

I have a question about Manual mesh leveling. I created a mesh, but I am not sure if its applied correctly. I know it applies by default before the print start, but I also applied in Cura slicer with the `M420 S1` command.

Does "Leveling active" checkbox need to be ON in order to apply the Mesh leveling? And what is the feature for?

And another question, if the mesh is applied, and while we are printing, shouldn't the Z coordinate live update/show position at the bottom of screen same as X and Y is updated while hotend is moving?

I guess it should, but the Z coordinate is static for every layer. And what I understood it should compensate (move up or down) because we applied mesh leveling.

Looking forward to hearing your answers.

PEace,

**Comment by Hologos** [comment]
I have the very same question. 🙏

**Reply by shtefcs** [reply]
If you don't have `M420 S1` command in your Slicer/Gcode, then you need to make it On everytime you start printer.

**Reply by Hologos** [reply]
I don't use M420 S1, I use G29 before every print. Do I still have this on? Thanks

**Reply by Tapap84** [reply]
M420 S1; Restore Manual Mesh

put that in your start gcode AFTER the homing line

**Reply by Hologos** [reply]
> M420 S1; Restore Manual Mesh
> 
> 
> 
> put that in your start gcode AFTER the homing line

Did you read the OP's question? Your answer doesn't explain anything.

**Reply by Tapap84** [reply]
let me try again. check box will turn itself on if you use m420 s1. as far as the z updating; it doesnt. my mesh is all within 0.08 of each other and the lcd will read 0.20 across the entire bed during first layer. it is active because i printed a 4 inch wheel and put it on my rod with a needle pointing out and you can see it turn as it moves across the bed following the mesh. so yes, the z will show static.

**Comment by shtefcs** [comment]
I don't know honestly

**Comment by Cal-LF** [comment, marked as answer]
It does not "have" to be on. The checkbox is there incase you need to turn it on as @shtefcs  stated.. It will not be on by default. Ref #1218.

**Reply by Chrisyahya** [reply]
so uh, M420 command is eliminating the need to tick the box in mesh leveling active?

**Reply by Cal-LF** [reply]
Yes.

**Reply by Chrisyahya** [reply]
thank you for your confirmation! gotta make that command default in cura.

**Comment by jawsawn** [comment]
I was going to open a new question but this is revelant. IF I turn on the checkbox do I not need to put M420 S1 on my gcode? Most of the videos I've seen informs about putting M420 S1 on the gcode but they are running an older version. Am I right to assume checking the box to make mesh leveling active injects the gcode itself or not?

**Reply by Jyers** [reply]
Checking the check box will enable the mesh, but will not be persistent on boot, so be careful if you decide to rely on it

**Comment by shtefcs** [comment]
>IF I turn on the checkbox do I not need to put M420 S1 on my code?

No you don't need it.

>Am I right to assume checking the box to make mesh leveling active injects the gcode itself or not?

Yes.

You either need one of those two options. Even if you have them both, it will not hurt.

**Reply by Gamma36** [reply]
It's a very confusing UI, because new users cannot predict the levelling.

Any option in the machine setting should overide GCode, to avoid chaos.

The situation is half the control comes from obscure Cura settings and half from the machine control panel. That's a recipe for complete disaster.

I've developed CAD/CAM software since 1981 and also found the Leveling section confusing, wasting a lot of time on failed initial layers.

**Comment by Rockument** [comment]
if Z coordinate not adjusting how do i know the BL/CR Touch has compensate a level. I tried to manual level as i could the CR touch probe do level measure and how do i know it actually works. cause while printing some of corners still unequal with the rest.

**Reply by Gamma36** [reply]
Theoretically the firmware is mapping the Cura workspace onto the printer workspace. It compensates Z moves to lay the initial layers onto the bed, with a consistent height. However it does not, as far as we know, change X and Y values, so there is a non-linear compensation.

The display is showing the GCode height, which is not precisely the bed height. As you say, ideally the display would show both the GCode height: 0, 0.2, 0.4, 0.6, 0.8 etc. and also the machine height above the bed like 0.03, 0.23, 0.43, 0.63, 0.83 etc. or perhaps the current Z offset like 0.03.


That lead to the bigger question, does the machine offset Z values for the entire 250mm height or does it slowly blend back to a flat Z-plane so that the final top layer is planar (perfectly flat)? A dev that worked on the mesh code could answer that question.

**Reply by Tapap84** [reply]
As far as the last part if I'm understanding you right, it does convert back to a flat plane. In the settings there something called fade height. I think it's set to 10. That means the mesh will slowly adjust to a flat plane over 10 layers so 11th should be flat.

**Reply by Gamma36** [reply]
@tititopher68-dev 
Thank you for the detailed info.

To explain myself better, the STL data lies somewhere within World space, it then gets transformed onto printer bed space. At this stage the model XYZ lie on a true workplane (point and 3 vectors in right handed orientation usually). The data is as it was designed in the CAD model.
A GCode printer toolpath is created, without any scaling or transformation.
(plastic injection moulding may apply a differential shrinkage to the model data at this stage - a nightmare to explain to customers!)

So the GCode is a 'rigid transform' of the CAD model.

Then the printer firmware does the mesh lookup offset to ensure the first N layers lie on the non-flat bed plate, by distorting Z values.

The user would benifit by seeing the Z distortion in real time, to verify the mesh is working and to know when the distortion ends.

Like @Rockument I was unsure if levelling was working. Feedback on printer Z changes (if any) would be very comforting. Then we could verify that the permutation of GCode commands and mesh correction were as expected.

Cheers
Paul

**Comment by Gamma36** [comment]
I suggest that when printing, if mesh levelling is active the Zoffset is drawn in RED and its value is the mesh modified Zoffset for the latest XY position.

Or drawn the Z value RED with its mesh modified height. 

When blending the mesh has stopped, 10 layers for example, revert to white for Z values.

**Comment by unixneo** [comment]
I am not an advanced user like most of the experts here, but personally, I prefer to use the "Leveling Active" checkbox in the UI and not M420 in the g-code. In this way, I can print the same g-code file with mesh leveling on or off (controlled by the printer), depending on the test case I am running. Maybe when I am more advanced I'll prefer not using the checkbox; but for now, I have found the checkbox invaluable and much appreciated for activating mesh leveling.

**Reply by Gamma36** [reply]
That's fine, just remember to set levelling in the UI and disable it in GCode. 

The advantage of GCode is you can automate levelling on every Print. For example UBL can quickly realign a saved mesh with 3 probed points.

**Reply by Gamma36** [reply]
@tititopher68-dev
One question, what is the difference between a valid and invalid mesh?
